src/overlays/actors/ovl_Obj_Syokudai/z_obj_syokudai.c

```diff
@@ -27,11 +29,243 @@ const ActorInit Obj_Syokudai_InitVars = {
+    s32 litTimeScale;
+    WaterBox* dummy;
+    f32 waterSurface;
+    s32 lightRadius = -1;
+    u8 lightColorY = 0;
+    Player* player;
+    EnArrow* arrow;
+    s32 itemType;
+    u32 toucherFlags;
+    Vec3f tipToFlame;
+    ColliderCylinder* collider1;
+    ColliderCylinder* collider2;
+
+    litTimeScale = torchCount;
+    if(torchCount == 10) {
+        torchCount = 24;
+    }
+    if (func_80042244(globalCtx, &globalCtx->colCtx, this->actor.posRot.pos.x, this->actor.posRot.pos.z, &waterSurface, &dummy) && ((waterSurface - this->actor.posRot.pos.y) > 52.0f )) {
+        this->timer = 0;
+        if (torchType == 1) {
+            Flags_UnsetSwitch(globalCtx, switchFlag);
+            if (torchCount != 0) {
+                this->timer = 1;
+            }
+        }
+    } else {
+        player = PLAYER;
+        itemType = 0;
+        if (this->actor.params & 0x400) {
+            this->timer = -1;
+        }
+        if (torchCount != 0) {
+            if (Flags_GetSwitch(globalCtx, switchFlag)) {
+                if (this->timer == 0) {
+                    this->timer = -1;
+                    if (torchType == 0) {
+                        func_80080480(globalCtx, &this->actor);
+                    }
+                } else if (this->timer > 0) {
+                    this->timer = -1;
+                }
+            } else if (this->timer < 0) {
+                this->timer = 20;
+            }
+        }
+        if (this->collider2.base.acFlags & 2) {
+            toucherFlags = this->collider2.body.acHitItem->toucher.flags;
+            if (toucherFlags & 0x20820) {
+                itemType = 1;
+            }
+        } else if (player->heldItemActionParam == 6) {
+            Math_Vec3f_Diff(&player->swordInfo[0].tip, &this->actor.posRot.pos, &tipToFlame);
+            tipToFlame.y -= 67.0f;
+            if ((SQ(tipToFlame.x) + SQ(tipToFlame.y) + SQ(tipToFlame.z)) < 400.0f) {
+                itemType = -1;
+            }
+        }
+        if (itemType != 0) {
+            if (this->timer != 0) {
+                if (itemType < 0) {
+                    if (player->unk_860 == 0) {
+                        player->unk_860 = 210;
+                        Audio_PlaySoundGeneral(NA_SE_EV_FLAME_IGNITION, &this->actor.projectedPos, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+                    } else if (player->unk_860 < 200) {
+                        player->unk_860 = 200;
+                    }
+                } else if (toucherFlags & 0x20) {
+                    arrow = (EnArrow*) this->collider2.base.ac;
+                    if(arrow->actor.update && (arrow->actor.id == 0x16)) {
+                        arrow->actor.params = 0;
+                        arrow->unk_1A8 = 0x800;
+                    }
+                }
+                if ((0 <= this->timer) && (this->timer < (50 * litTimeScale + 100)) && (torchType != 0)) {
+                    this->timer = 50 * litTimeScale + 100;
```

Comment:
so `timer` is `litTimer`? (the time before the torch isn't lit anymore)
I can't tell what's going on in this `if` jungle

Reply:
I can add some comments, but it's basically doing all the checks of the torch interacting with relevant items (deku sticks, arrows, fire attacks), lighting the torch if unlit and setting its lit timer, and checking if enough torches are lit to solve the puzzle.